core/src/type_params.rs

```diff
@@ -0,0 +1,307 @@
+        }
+    }
+}
+
+uses_type_params!(syn::Variant, fields);
+uses_type_params!(syn::Field, ty);
+uses_type_params!(syn::DataStruct, fields);
+uses_type_params!(syn::AngleBracketedGenericArguments, args);
+uses_type_params!(syn::BareFnArg, ty);
+uses_type_params!(syn::QSelf, ty);
+uses_type_params!(syn::TypePath, qself, path);
+uses_type_params!(syn::TypeBareFn, inputs, output);
+uses_type_params!(syn::ParenthesizedGenericArguments, inputs, output);
+uses_type_params!(syn::TraitBound, path);
+uses_type_params!(syn::TypeTraitObject, bounds);
+uses_type_params!(syn::TypeImplTrait, bounds);
+uses_type_params!(syn::Binding, ty);
+
+impl UsesTypeParams for syn::DataEnum {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        self.variants.collect_type_params(type_set)
+    }
+}
+
+impl UsesTypeParams for syn::Fields {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        self.collect_type_params(type_set)
+    }
+}
+
+/// Check if an Ident exactly matches one of the sought-after type parameters.
+impl UsesTypeParams for Ident {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        type_set.iter().filter(|v| *v == self).collect()
+    }
+}
+
+impl UsesTypeParams for syn::ReturnType {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        if let syn::ReturnType::Type(_, ref ty) = *self {
+            ty.uses_type_params(type_set)
+        } else {
+            Default::default()
+        }
+    }
+}
+
+impl UsesTypeParams for Type {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        match *self {
+            Type::Slice(ref v) => v.elem.uses_type_params(type_set),
+            Type::Array(ref v) => v.elem.uses_type_params(type_set),
+            Type::Ptr(ref v) => v.elem.uses_type_params(type_set),
+            Type::Reference(ref v) => v.elem.uses_type_params(type_set),
+            Type::BareFn(ref v) => v.uses_type_params(type_set),
+            Type::Tuple(ref v) => v.uses_type_params(type_set),
+            Type::Path(ref v) => v.uses_type_params(type_set),
+            Type::Paren(ref v) => v.elem.uses_type_params(type_set),
+            Type::Group(ref v) => v.elem.uses_type_params(type_set),
+            Type::TraitObject(ref v) => v.uses_type_params(type_set),
+            Type::ImplTrait(ref v) => v.uses_type_params(type_set),
+            Type::Macro(_) | Type::Verbatim(_) | Type::Infer(_) | Type::Never(_) => {
+                Default::default()
+            }
+        }
+    }
+}
+
+uses_type_params!(syn::TypeTuple, elems);
+
+impl UsesTypeParams for syn::Path {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        // Not sure if this is even possible, but a path with no segments definitely
+        // can't use type parameters.
+        // We don't use `is_empty` here because that also considers punctuation.
```

Comment:
It's not possible for a Punctuated to contain only punctuation.

Reply:
Thanks, I'll go ahead and clean that up to just use `is_empty` then.